File: 0169-majority-element/0169-majority-element.cpp
// Given: array nums, size = n
// number of the Majority element > n/2

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        // Create a hash map to store the count of each number
        unordered_map<int, int> table;
        int n = nums.size();

        // Iterate through the array and count each numbers's frequency
        for(int num : nums) {
            // Increment count for the current number
            table[num] += 1; 

            // if the count of current num exceeds n/2, return it
            if (table[num] > n/2) 
                return num;
        }

        // Return -1 as a fallback
        return -1;
    }
};

File: 0169-majority-element/0169-majority-element.js
/**
 * @param {number[]} nums 
 * @return {number} 
 * majority element always exists (appears more than [n/2] times)
 * constraints: majority elem always exists (# of majority = always one)
 * 
 * Boyer–Moore Voting Algorithm
 * - Maintain a candidate and a counter
 * - Increase count for same value, decrease for different
 * - Majority element survives all cancellations
 */

var majorityElement = function(nums) {
    let candidate;
    let count = 0;

    for (const num of nums) {
        // choose a candidate: if count is 0, set the candidate as the current num
        if (count === 0) candidate = num;
        // if candidate is same as num, increase a count
        if (candidate === num) count++;
        // if different, decrease a count
        else count--;
    }

    // candidate is guaranteed to be the majority element
    return candidate;
};